### ISSUE TITLE
[📍] Customizable repeat_count for no-metric tryjobs

### ISSUE THREAD

**Issue by dave-2** [catapult-project/catapult#4332]
Currently, Pinpoint runs 10 repeats for try jobs. Sometimes, people want a different repeat count.

For example, someone ran a [job](https://pinpoint-dot-chromeperf.appspot.com/job/16ee436e440000) on `system_health.common_desktop` with no story, which takes 45 minutes to run. Currently, that's 10 repeats without the patch and 10 repeats with, so 15 hours of bot time in one job.

We should default to 1 to save bot resources, but let users change it in the dialog, in case they want more data and know what they are doing.

@anniesullie @simonhatch @perezju

**Comment by perezju** [comment]
I thought about these kinds of things when running a job for [crbug.com/804324](http://crbug.com/804324), where [the results](https://pinpoint-dot-chromeperf.appspot.com/results2/17eb2426440000) came back showing some "large" relative changes (from 22% improvements to nearly 28% regressions) however, according to the :neutral_face: shown next to the results none of the changes are in fact statistically significant. So this leaves me a bit unsure of what to make of the results: is the CL really affecting the metrics of that benchmark or not really?

On the other hand, this is precisely the kind of question that pinpoint in bisect mode is asking when comparing each pair of revisions: did the distribution of values before/after changed? And that is working wonderfully.

From the try-job side, however, the number of 10 repeats is pretty ad-hoc. I would have wanted instead for pinpoint to run enough times to figure out which (if any) of the benchmark metrics had changed significantly with/without the patch.

I understand there is the question of resources, and how we make this scale up, but this a question I would really like to be able to ask pinpoint: this is my CL, what effect is it going to have on all platforms/benchmarks? (without having to actually land the CL, which is what we do today)

**Comment by anniesullie** [comment]
@perezju that is a good point!

I would like to hear what @pasko thinks, since we've gotten questions from him about tuning perf try job parameters in the past.

**Comment by dave-2** [comment]
Oh, so what we really want is for Pinpoint to determine the repeat counts by using its bisection comparison across all the histograms.

**Comment by simonhatch** [comment]
Another use case from Benchmarking team is using Pinpoint to determine if benchmarks or changes to them are flaky/noisy.

**Comment by pasko** [comment]
Thanks for keeping me in the loop :)

### TL;DR:
1. we have reasonable ways to set repeat_count at benchmark level already
2. selecting more bisect time parameters at benchmark level is desirable (like statistical method, various cutoff thresholds etc.) 
3. local bisects for complex cases would be needed once in a while, please make it not too difficult :)

### Longer version:

I think 10 is quite a reasonable tradeoff for statistically sound comparisons and bot resources.

The problem with "run enough times to figure out which (if any) of the benchmark metrics had changed significantly" is that it is against the general "philosophy" of statistical analysis. Statistical methods are made to take "as much data as you can afford" and making the analysis once (with some allowance to make a mistake). They do so to limit the effect of "type 1 error" (like: by massaging the data for long enough any patient would be proven to have a disease when in fact the patient does not have the disease). Deviating from this philosophy will need some research ...

I think setting default repeat_count per-benchmark is currently the best approach. In overwhelming majority of cases perf sheriffs must **not** tweak repeat_count when they launch bisects because they are unaware of noise levels, probability distributions of the metric, time to run the bisect, likelihood of finding the culprit, etc. All these factors need to be taken into consideration. This puts the benchmark owners in the best position to make the call (IMHO). It also allows trial-n-error and ability to learn the properties of the benchmark over time.

Regardless of what is best in overwhelming majority of cases, sometimes we have to do off-band analysis (new features, looking for insights, complex cases that Pinpoint does not cover yet, etc.), where we can afford unusually long running times (by running locally), and would need to tweak repeat_count and a number of other parameters. For this case local bisects would be best, but given that it would be used rarely, every time it is used, it will likely have slightly rotten. I'm OK with that, but not sure how much happier we would be with that overall. I imagine something like providing pinpoint algorithm as a library. We would use it with plain git-bisect or something similar (at experimenters' own risk).
